obelisk/cpp/obelisk_cpp/include/obelisk_node.h

```diff
@@ -0,0 +1,820 @@
+
+            CreatePublishers();
+            CreateSubscriptions();
+            CreateTimers();
+
+            RCLCPP_INFO_STREAM(this->get_logger(), this->get_name() << " configured.");
+            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+        }
+
+        /**
+         * @brief activates up the node.
+         *
+         * @param prev_state the state of the ros node.
+         */
+        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn virtual on_activate(
+            const rclcpp_lifecycle::State& prev_state) {
+            rclcpp_lifecycle::LifecycleNode::on_activate(prev_state);
+
+            for (auto& [key, pub] : publishers_) {
+                if (pub) {
+                    pub->OnActivate();
+                }
+            }
+
+            for (auto& [key, timer] : timers_) {
+                if (timer) {
+                    timer->reset(); // Start the timer
+                }
+            }
+
+            RCLCPP_INFO_STREAM(this->get_logger(), this->get_name() << " activated.");
+            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+        }
+
+        /**
+         * @brief deactivates up the node.
+         *
+         * @param prev_state the state of the ros node.
+         */
+        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn virtual on_deactivate(
+            const rclcpp_lifecycle::State& prev_state) {
+            rclcpp_lifecycle::LifecycleNode::on_deactivate(prev_state);
+
+            for (auto& [key, pub] : publishers_) {
+                if (pub) {
+                    pub->OnDeactivate();
+                }
+            }
+
+            for (auto& [key, timer] : timers_) {
+                if (timer) {
+                    timer->cancel(); // Stop the timer
+                }
+            }
+
+            RCLCPP_INFO_STREAM(this->get_logger(), this->get_name() << " deactivated.");
+            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+        }
+
+        /**
+         * @brief cleans up the node.
+         *
+         * @param prev_state the state of the ros node.
+         */
+        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn virtual on_cleanup(
+            const rclcpp_lifecycle::State& prev_state) {
+            rclcpp_lifecycle::LifecycleNode::on_cleanup(prev_state);
+
+            // Clear current data
+            callback_groups_.clear();
+
+            // Clean up publishers
+            for (auto& [key, pub] : publishers_) {
+                if (pub) {
+                    pub->Release();
+                }
+            }
+            publishers_.clear();
+            registered_publishers_
+                .clear(); // TODO: Consider if this should be called, or if I should leave this in place
```

Comment:
discuss todo over slack to clarify decision

Reply:
I'm gonna leave this here I think. The only reason to remove it is if we want to let the user shutdown or cleanup and then bring back up the node without registering all the components again. But imo, that is the point of the shutting down: that the node is done. So ill leave clearing this here.